### ISSUE TITLE
dyld: initializer function 0x0 not in mapped image for /usr/local/bin/node

### ISSUE THREAD

**Issue by qiulang** [tj/n#634]
# Bug Report

## Summary

When network is unstable I keep hitting this error. When it happened n will also **break my installed nodejs**!  

~ ➤ sudo n lts

  installing : node-v14.15.1
       mkdir : /usr/local/n/versions/node/14.15.1
       fetch : https://nodejs.org/dist/v14.15.1/node-v14.15.1-darwin-x64.tar.xz
######################################################                                                                        44.2%
curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
bin/node: Lzma library error:  No progress is possible
dyld: initializer function 0x0 not in mapped image for /usr/local/bin/node

dyld: initializer function 0x0 not in mapped image for /usr/local/bin/node

   installed :  (with npm )

~ ➤ node -v
dyld: initializer function 0x0 not in mapped image for /usr/local/bin/node

```bash
~ ➤ command -v node
/usr/local/bin/node
~ ➤ n --version
6.7.0
~ ➤ node -p process.platform
darwin
```

**Comment by shadowspawn** [comment]
It looks like the download of the archive fails, but tries to expand anyway which fails, but tries to install anyway which has a bad result! I'll look into the error handling.

As you probably found, once you have a bad download that version will get installed from cache and stay broken. You can delete it from the cache to try the download again:

```
n rm 14.15.1
```

**Comment by qiulang** [comment]
Yes I have found that I need to delete the download files, but the problem is my node was broken then so I need to reinstall it first (And that was my issue, it should n't corrupt my node even the download failed.).

**Comment by qiulang** [comment]
BTW, it is possible to let n switch back to node version pre-install before n ? For example like `nodenv local --unset` ?

**Comment by shadowspawn** [comment]
> BTW, it is possible to let n switch back to node version installed before n ? 

Only if it is installed to a different location. See `n uninstall` described under: https://github.com/tj/n#removing-versions

**Comment by qiulang** [comment]
So does that mean I need to uninstall all the node versions installed by n to revert to the system version of node ?

**Comment by shadowspawn** [comment]
No. `n` keeps copies of the downloaded versions you have installed before so it can install them again without needing to download them. When you install, everything is copied from the cache (`/usr/local/n`) to the install locations (e.g. `/usr/local/bin/node`).

`n rm 1.2.3` deletes the cached version, it does not affect the installed version.

`n uninstall` deletes the installed version, it does not affect the cached versions.

**Comment by qiulang** [comment]
Hi I tried `n uninstall `and I found it is quite "inconvenient" compare to `nodenv local --unset` because when I install node from node installer it will install node in /usr/local/bin, so `n uninstall` can't revert to the system version of node in my case (mac).
If I open a feature request will you consider it ?

**Comment by shadowspawn** [comment]
You can configure the root directory where `n` installs node and caches the downloads using `N_PREFIX`. A common use case is to install to a folder in your home folder.

**Comment by shadowspawn** [comment]
I have reproduced the install failure by adding `--max-time 3` to the curl download. (Also using Mac.)

**Comment by qiulang** [comment]
@shadowspawn Do you have a schedule to fix it ? Is it quite annoying that whenever I have a bad connection my nodejs will be broken and I have to reinstall it again.

Thanks!

**Comment by shadowspawn** [comment]
I didn't have a schedule, but I'll try and do a release within next week since you asked! 😄

**Comment by shadowspawn** [comment]
Improvement released in v6.7.1.
